src/ghost_borrow_mut.rs

```diff
@@ -247,6 +246,63 @@ generate_public_instance!(a, b, c, d, e, f, g, h, i, j, k, l ; T0, T1, T2, T3, T
 //  Implementation
 //
 
+//  This trait is a pure implementation detail, which only needs to be public because it must be part of the signature
+//  of public functions/traits.
+mod sealed {
+
+#[derive(Clone, Copy, Debug)]
+pub enum Span {
+    Single(*const u8),
+    Slice(*const u8, *const u8),
+}
+
+#[doc(hidden)]
+pub trait GetSpan {
+    //  #   Safety
+    //
+    //  At the moment, it is assumed that the address of a fat pointer always points to the first byte of the actual data.
+    //  If this does not hold, the span returned will not help enforce the necessary invariants.
+    fn get_span(&self) -> Span;
+}
+
+impl<T: ?Sized> GetSpan for T {
+    default fn get_span(&self) -> Span { Span::Single(self as *const T as *const u8) }
+}
+
+impl<T> GetSpan for [T] {
+    fn get_span(&self) -> Span {
+        if self.len() <= 1 {
+            Span::Single(self.as_ptr() as *const u8)
+        } else {
+            let start = self.as_ptr() as *const u8;
+            let length = self.len() * core::mem::size_of::<T>();
+
+            //  Safety:
+            //  -   `start` is a valid pointer.
+            //  -   `start + length` points 1 byte past the end of a valid slice.
+            Span::Slice(start, unsafe { start.add(length) })
```

Comment:
I'm not sure if the pointer being 1 past the end of the object ensures that it's not overflowing. Can it be at the exact end of memory?

Reply:
`core::slice::Iter` does not seem to worry about that: https://doc.rust-lang.org/src/core/slice/iter.rs.html#88.